File: Difficulty: Medium/Strings Rotations of Each Other/strings-rotations-of-each-other.py
class Solution:
    def areRotations(self, s1, s2):
        # code here
        s1=s1+s1
        res= s2 + '#' + s1
        n= len(res)
        z=[0]*n
        l=0
        r=-1
        m=len(s2)
        for i in range(1,n):
            
            if i<=r:
                z[i]=min(r-i+1,z[i-l])
            while i+z[i]<n and res[z[i]]==res[i+z[i]]:
                z[i]+=1
            
            if (i+z[i]-1) > r:
                r=i+z[i]-1
                l=i
                
                    
        
        
        for i in range(m+1,n):
            if z[i]==m:
                return True
        
        return False

File: Difficulty: Medium/Strings Rotations of Each Other/strings-rotations-of-each-other.cpp
class Solution {
public:
    bool areRotations(string &s1, string &s2) {
        if (s1.size() != s2.size()) return false;
        if (s1.empty()) return true;

        string ss = s1 + s1;                
        string res = s2 + '#' + ss;          
        int n = res.size();
        vector<int> z(n, 0);
        int l = 0, r = 0;

        int req = (int)s2.size();
        for (int i = 1; i < n; ++i) {
            if (i <= r) z[i] = min(r - i + 1, z[i - l]);
            while (i + z[i] < n && res[z[i]] == res[i + z[i]]) ++z[i];
            if (i + z[i] - 1 > r) { l = i; r = i + z[i] - 1; }
            if (z[i] == req) return true;
        }
        return false;
    }
};
